Convert color adjust types to strongly typed CSS/Style values

weinig opened the pull request:
#### 6baedaebad6df0ea330b0f250355b9135395bc74
<pre>
Convert color adjust types to strongly typed CSS/Style values
<a href="https://bugs.webkit.org/show_bug.cgi?id=282124">https://bugs.webkit.org/show_bug.cgi?id=282124</a>

Reviewed by Darin Adler.

Converts the `color-scheme` property from CSS Color Adjust
to the new strongly typed CSS/Style values and updates the
parsing and serialization to match the spec.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color-adjust/parsing/color-scheme-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color-adjust/parsing/color-scheme-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color-adjust/parsing/color-scheme-valid-expected.txt:
    - Update now passing tests.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add new files.

* Source/WebCore/css/values/color-adjust/CSSColorScheme.cpp: Added.
* Source/WebCore/css/values/color-adjust/CSSColorScheme.h: Added.
* Source/WebCore/style/values/color-adjust/StyleColorScheme.cpp: Added.
* Source/WebCore/style/values/color-adjust/StyleColorScheme.h: Added.
    - Add new types.

* Source/WebCore/css/CSSColorSchemeValue.cpp: Added.
* Source/WebCore/css/CSSColorSchemeValue.h: Added.
* Source/WebCore/css/CSSValue.cpp:
* Source/WebCore/css/CSSValue.h:
    - Added CSSValue wrapper for new CSS::ColorScheme value.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
    - Call through to new toCSS implementation.

* Source/WebCore/style/StyleBuilderConverter.h:
    - Call through to new toStyle implementation.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+ColorAdjust.cpp: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ColorAdjust.h: Added.
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/css/process-css-properties.py:
    - Move parsing of color-scheme to spec specific file and re-implement to match
      current spec.

* Source/WebCore/css/values/CSSValueTypes.cpp: Added.
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/style/values/StyleValueTypes.h:
    - Add new shared CSS/Style::CustomIdentifier type that wraps AtomString.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
    - Update to use new Style::ColorScheme type.

* Source/WebCore/rendering/style/StyleColorScheme.h: Removed.
    - Remove old StyleColorScheme type that is no longer used.

Canonical link: <a href="https://commits.webkit.org/286549@main">https://commits.webkit.org/286549@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/beaea18efd1c440d2ce5f9059306023115a92780

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76309 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55339 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29211 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/80820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/27581 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78425 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64483 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3635 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/80820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/27581 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79376 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/49739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/65534 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/80820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/47137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/23021 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/25903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/68264 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/23355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82276 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3681 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/2407 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/68058 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3835 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/65506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67370 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16793 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/11336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/9436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3629 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6436 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/3652 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/7081 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5410 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/css/calc/CSSCalcTree+Copy.h:
```diff
@@ -29,11 +29,12 @@
 Anchor::Side copy(const Anchor::Side&);
 
+// FIXME: This is exposed for copyAndSimplifyChildren for Indirect<Anchor>. Find a better solution.
+Anchor::Side copy(const Anchor::Side&);
```
This appears to be the exact same declaration as line 34?

weinig replied:
yeah, merge fun! Will remove.

darinadler commented on LayoutTests/css-dark-mode/color-scheme-css.html:
```diff
@@ -148,8 +148,8 @@
-    test_prop(document.body, "color-scheme", "dark");
-}, "Color schemes changed to a bogus value and dark");
+    test_prop(document.body, "color-scheme", "foo dark");
+}, "Color schemes changed to a unknown value and dark");
```
an unknown

darinadler commented on LayoutTests/css-dark-mode/older-syntax/supported-color-schemes-css.html:
```diff
@@ -93,9 +93,9 @@
-}, "Supported color schemes changed to light and a bogus value");
+    test_prop(document.body, "supported-color-schemes", "light foo");
+    test_prop(document.body, "color-scheme", "light foo");
+}, "Supported color schemes changed to light and a unknown value");
```
an unknown

darinadler commented on LayoutTests/css-dark-mode/older-syntax/supported-color-schemes-css.html:
```diff
@@ -154,9 +154,9 @@
-    test_prop(document.body, "color-scheme", "dark");
-}, "Supported color schemes changed to a bogus value and dark");
+    test_prop(document.body, "supported-color-schemes", "foo dark");
+    test_prop(document.body, "color-scheme", "foo dark");
```
an unknown

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+ColorAdjust.cpp:
```diff
@@ -0,0 +1,119 @@
+    if (range.peek().id() == CSSValueNormal) {
+        range.consumeIncludingWhitespace();
+
+        // NOTE: `normal`, is represented in CSS::ColorScheme as an empty list of schemes.
```
Remove the stray comma here?

darinadler commented on Source/WebCore/css/values/CSSValueTypes.h:
```diff
@@ -45,13 +47,21 @@ namespace CSS {
+    AtomString value;
+
+    bool operator==(const Identifier&) const = default;
+    bool operator==(const AtomString& other) const { return value == other; }
```
Would this be better as a pair of non-member functions, to allow comparisons in either order?

weinig replied:
In c++20, this is no longer needed! https://brevzin.github.io/c++/2019/07/28/comparisons-cpp20/. The member version works for both orders now (and also works for !=).

darinadler replied:
How did I not remember that? I remembered the `!=` goodness, which is why I didn’t bring that up.

darinadler commented on Source/WebCore/css/values/CSSValueTypes.h:
```diff
@@ -784,6 +818,15 @@ template<typename CSSType> struct CSSValueChildrenVisitor<RectEdges<CSSType>> {
 
+namespace WTF {
+
+inline TextStream& operator<<(TextStream& ts, const WebCore::CSS::Identifier& value)
```
To me it seems excessive to inline this; doesn’t seem so performance critical to me. Would be nice to not have to include `TextStream.h` in this header.

darinadler commented on Source/WebCore/css/values/CSSValueTypes.h:
```diff
@@ -358,6 +368,14 @@ template<CSSValueID C> struct Serialize<Constant<C>> {
 
+// Specialization for `Identifier`.
+template<> struct Serialize<Identifier> {
+    void operator()(StringBuilder& builder, const Identifier& value)
```
Does this really need to be inlined? Not sure about compile time tradeoffs.

weinig replied:
It does not. I am so used to this file being about template types, the first one that was not didn't cross my mind to be out of line.

darinadler commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -4859,7 +4859,7 @@ void Document::metaElementColorSchemeChanged()
     RefPtr<CSSValue> colorScheme;
     auto colorSchemeString = emptyString();
-    auto cssParserContext = CSSParserContext(document());
+    const auto& cssParserContext = this->cssParserContext();
```
I’d be tempted to use a shorter name in this small function. Like just `context`.

darinadler commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9116,12 +9116,12 @@ bool Document::useDarkAppearance(const RenderStyle* style) const
-    // Fallback to the document's supported color schemes if style was empty (auto).
-    if (colorScheme.isEmpty())
+    if (style) {
+        if (style->colorScheme().schemes.isEmpty())
```
```
bool isNormal = !style || style->colorScheme().isNormal();
auto colorScheme = isNormal ? m_colorScheme : style->colorScheme().colorScheme();
```

weinig replied:
Nice.

weinig committed:
Convert color adjust types to strongly typed CSS/Style values
https://bugs.webkit.org/show_bug.cgi?id=282124

Reviewed by Darin Adler.

Converts the `color-scheme` property from CSS Color Adjust
to the new strongly typed CSS/Style values and updates the
parsing and serialization to match the spec.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color-adjust/parsing/color-scheme-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color-adjust/parsing/color-scheme-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color-adjust/parsing/color-scheme-valid-expected.txt:
    - Update now passing tests.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add new files.

* Source/WebCore/css/values/color-adjust/CSSColorScheme.cpp: Added.
* Source/WebCore/css/values/color-adjust/CSSColorScheme.h: Added.
* Source/WebCore/style/values/color-adjust/StyleColorScheme.cpp: Added.
* Source/WebCore/style/values/color-adjust/StyleColorScheme.h: Added.
    - Add new types.

* Source/WebCore/css/CSSColorSchemeValue.cpp: Added.
* Source/WebCore/css/CSSColorSchemeValue.h: Added.
* Source/WebCore/css/CSSValue.cpp:
* Source/WebCore/css/CSSValue.h:
    - Added CSSValue wrapper for new CSS::ColorScheme value.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
    - Call through to new toCSS implementation.

* Source/WebCore/style/StyleBuilderConverter.h:
    - Call through to new toStyle implementation.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+ColorAdjust.cpp: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ColorAdjust.h: Added.
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/css/process-css-properties.py:
    - Move parsing of color-scheme to spec specific file and re-implement to match
      current spec.

* Source/WebCore/css/values/CSSValueTypes.cpp: Added.
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/style/values/StyleValueTypes.h:
    - Add new shared CSS/Style::CustomIdentifier type that wraps AtomString.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
    - Update to use new Style::ColorScheme type.

* Source/WebCore/rendering/style/StyleColorScheme.h: Removed.
    - Remove old StyleColorScheme type that is no longer used.

Canonical link: https://commits.webkit.org/286549@main
